### ISSUE TITLE
@override and super

### ISSUE THREAD

**Issue by mrimvo** [godotengine/godot-proposals#3936]
### Describe the project you are working on

A game prototype in Godot 4

### Describe the problem or limitation you are having in your project

I often forget to call super()

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Goals we want to achieve
1. All methods are **treated the same**
    * make no difference between engine-called and other methods.
2. **Keep the simplicity** of Godot 3's functions
    * no need to call super, eg in _ready()

Suggestion to achive this goal: introduce the new annotation `@override`. As the name says, if you use this keyword, the method is actually *overriden*, and the super function is *not* called, except you explicitely do so. This is what you would expect in most OOP languages. If you omit `@override`, `super` is automatically called.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Let's see some code examples:

```
 # good old Godot 3 way
 # .super() is called implicitly
 # this is nice because beginners always forget to call super.
 # it also is easier to read and write
func _ready():
    do_something()
```

```
 # this is perfectly possible, but it's the same as the example before
override func _ready():
     super()
     do_something()
```

 ```
# using super without override
func _ready():
    do_something()
    super()  # produces an ERROR: cannot call super without override
```

```
 # we use override because we want full control when to call super()
override func _ready():
    do_something()
    if condition:
        super()
```

```
 # we don't want the super function to be called
override func _ready():
    do_something()
```

# Conclusion

Using `@override` and `super` we **make easy things easy and hard things possible**. The majority of functions won't need `override` nor `super`, as in most cases you want the super function to be called. This makes for **more readable and clean code**, as you won't have to explicitly call *super* in most cases.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This enhancement will be used very often, it affects all functions.

### Is there a reason why this should be core and not an add-on in the asset library?

It's a GDScript feature

**Comment by mrimvo** [comment]
This Proposal presents an alternative Proposal to #1631. It is no duplicate.

**Comment by Mickeon** [comment]
Since Godot 4 removed the implicit `super()` call on `_ready()`, `_process()` and `_physics_process()`, it really does feel repetitive to state it every time. So, I concur on this, but as this comment states:
```
 # this is nice because beginners always forget to call super.
```
I'm afraid the same could be said for a new `@override` annotation. I've read of users not realising that `_ready()` [implicitly called super()](https://godotengine.org/qa/9244/can-override-the-_ready-and-_process-functions-child-classes).
Perhaps with improvements to the documentation it could easily be mitigated.
A cookie-cut note along the lines of "This method implicitly calls `super()`. To prevent this, use the `@override` annotation" could do.

**Comment by zinnschlag** [comment]
Good idea. Its clean, convenient and consistent, but without losing flexibility. And as an additional benefit it is compatible with Godot 3 in many cases, which will reduce the need for porting when going 3 to 4.

Just one nitpick: You state as target version Godot 4.x, which could mean any Godot 4 version, but usually means 4.1 or later. That's not possible, since it would break compatibility between 4.0 and later versions. So it really should be 4.0.

**Comment by dalexeev** [comment]
For void functions, this seems like a good idea (although I still have some doubts). But you forgot about non-void functions. In this case, it will not be possible to make automatic behavior, because a function can only return one value; an explicit `super` call is required.

**Comment by mrimvo** [comment]
> I've read of users being puzzled because _ready() [implicitly called super()](https://godotengine.org/qa/9244/can-override-the-_ready-and-_process-functions-child-classes).

I agree @Mickeon. I think much of the confusion in Godot 3 was caused because some functions implicitly called super, and some not.

@zinnschlag I stumbled over this issue when I migrated code from Godot 3 to 4: I had to insert super() in many places. Thanks for the hint about the target version, I edited the original post.

@dalexeev I see, true these examples are missing. I'm going to add them to the original post.

**Comment by mrimvo** [comment]
@dalexeev I see, true these examples were missing, I added those to the original post. In case of non-void functions you always got two options: to dismiss the value of the super function, or keep it. In case you *must* call super, the IDE would tell you to use `@override`, see OP. I think your use case works really well with the proposed concept. Please feel free to challenge this concept more and express your doubts, so we can see if this idea covers all cases we can think of.

**Comment by cogspace** [comment]
I ran into a problem the other day with `_enter_world()` which I didn't even know existed. I overrode it on accident. Mandatory `@override` would save a lot of headache in these cases.

**Comment by mrimvo** [comment]
The first Godot 4 Beta having released recently, I would like to bring this ticket back to attention.

As @zinnschlag pointed out, if the proposed behavior is desired, it must be implemented in 4.0, as it breaks backwards compatibility.

I would like to make sure this ticket is brought to the attention of the maintainer(s) of GDScript, so they can make a conscious decision if this concept should be part of 4.0 or not.

To underline the importance of this concept, GDScript and Inheritance are two very powerful concepts in Godot and having a rock solid and user-friendly mechanism to override methods is key. We don't want boilerplate super calls and at the same time we want GDScript to be beginner friendly. Forgetting to call super, or accidentally overriding methods are some of the most annoying bugs to keep hold on, as currently nothing actively keeps you from doing that, and the behavior silently fails. These issues are addressed with the proposed concept.

**Comment by dalexeev** [comment]
For me, it would be acceptable if `@override` was required in the child class, and the behavior of `super` remained current.

I still think implicitly calling a parent method in the absence of `@override` is very bad. On the contrary, we are trying to get away from implicit behavior and that is why we removed multilevel calls.

We could also add `@abstract`/`abstact` (you must define all abstract methods in the child class so that the child class is non-abstract) and `@virtual`/`virtual` (you can define or not define a virtual method in descendant classes, but before that the method is considered undefined).

**Comment by YuriSizov** [comment]
> I would like to make sure this ticket is brought to the attention of the maintainer(s) of GDScript, so they can make a conscious decision if this concept should be part of 4.0 or not.

GDScript is not going to receive any changes for 4.0, as we are satisfied with its feature set and want to make it stable. Same applies to all of the engine at this point, but I wanted to clarify that because GDScript has been in a feature-freeze mode even before the beta stage.

**Comment by p10tr3k** [comment]
Currently (Godot 4.0.2) when you create method `_ready` you do not know if you are overriding some code. Thus you are not sure if you should call `super()` or not. I would be great to use `@override` keyword only when this method is implemented already in inheritance chain. If it is not implemented yet but is virtual, you will not have to use `@override` keyword.

**Comment by dalexeev** [comment]
> Thus you are not sure if you should call `super()` or not. I would be great to use `@override` keyword only when this method is implemented already in inheritance chain.

This makes sense, but for many users it still be inconvenient for custom methods (when you override the "abstract" method of the base script). This can be resolved if we also add the `abstract`/`virtual` annotation/keyword.

Another possible solution would be to add a `super(params)` comment when autocompleting the method override in the editor. See https://github.com/godotengine/godot-proposals/discussions/6639#discussioncomment-5538597 and https://github.com/godotengine/godot/pull/75013.

**Comment by Zireael07** [comment]
That comment is a good idea indeed <3

**Comment by sketchyfun** [comment]
This keeps tripping me up while porting my game to 4.x, so I'd love to see something like this implemented

**Comment by Calinou** [comment]
@sketchyfun Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by SirPigeonz** [comment]
TIP if you don't want to call super in `_ready()`:

```gdscript
@onready var foil = wrapper()


func wrapper():
# Code that you want to be run on ready in every class extending this one.
```

**Comment by rshouker** [comment]
the new @override in python prevents accidental typos in methods, or a mistake in method signature (parameter types) when trying to override. Without it a new method would be created that does not overrides that of the base (super) class.
For me this is the major benefit of @override.

**Comment by Mickeon** [comment]
Please enclose the annotation with \`, like \`@override\`. Without it you mention a random user on GitHub.

**Comment by pompyproductions** [comment]
I feel like having an `@override` would work particularly well (at least for my approach) with resources that define custom behavior. Having the editor throw a _"The method definition is already implemented in the base class. If this is intended, use `@override`."_ could nip many a bug in the bud. Imagine a "deal_damage" effect that ignores enemy defense, status effects with a particular stat check (combined with super), etc etc...

Editor messages help beginners like me a ton, since they give meaningful insight to a potential problem, as opposed to things simply not working as intended, and having to re-read through every single piece of code. Would love to see this happen! 

(I don't post much so please let me know if my comment does not contribute particularly to the topic at hand.)

**Comment by danijmn** [comment]
> I feel like having an `@override` would work particularly well (at least for my approach) with resources that define custom behavior. Having the editor throw a _"The method definition is already implemented in the base class. If this is intended, use `@override`."_ could nip many a bug in the bud. Imagine a "deal_damage" effect that ignores enemy defense, status effects with a particular stat check (combined with super), etc etc...
> 
> Editor messages help beginners like me a ton, since they give meaningful insight to a potential problem, as opposed to things simply not working as intended, and having to re-read through every single piece of code. Would love to see this happen!
> 
> (I don't post much so please let me know if my comment does not contribute particularly to the topic at hand.)

I totally agree. Until there is a consensus on what should be done from a long-term perspective, and to provide a simple solution that's (hopefully) easy to implement and covers the most important aspect (i.e. preventing mistakes), I think at least **there should be an editor option for raising a warning when shadowing a function from an ancestor class, as in most cases this is probably not intentional**.

As it stands, in most cases the practical consequence of not having a warning is that the implementer forgets to call super(), probably resulting in unexpected behavior. This is even more likely to happen in lifecycle methods such as _ready() and _update() than in custom named-methods, as the former are used universally across the project and it's easy to forget that they are shadowing lifecycle methods in parent classes.

IMO such a warning should be enabled by default, but configurable under `Project Settings > General > Debug > GDScript` as usual.

**Comment by Leonardo-Ma** [comment]
While we unfortunately don't have this,

We have `@abstract`, that allows:

```gdscript
## Generic abstract class for entities that take damage and attack...
@abstract class_name AggressiveEntity
extends CharacterBody3D
```
And it can have:
```gdscript
## Virtual method for subclasses to override instead of _ready()
@abstract func _child_ready() -> void
```
Forcing it to be implemented and reminding that **I should not override the _ready()** in this child
